job-runner/chain/reaper.go

```diff
@@ -0,0 +1,589 @@
+	// parent jobs, it will get jobData copied from all of its parent jobs in
+	// the order that they finish. Therefore, if multiple parent jobs try to set
+	// the same jobData field, it is undetermined which parent job the job will
+	// receive the final copy of that jobData field from.
+	for _, nextJob := range r.chain.NextJobs(job.Id) {
+		nextJLogger := jLogger.WithFields(log.Fields{"next_job_id": nextJob.Id})
+
+		for k, v := range job.Data {
+			nextJob.Data[k] = v
+		}
+		r.chainRepo.Set(r.chain)
+
+		// Enqueue any runnable child jobs.
+		if !r.chain.IsRunnable(nextJob.Id) {
+			nextJLogger.Infof("next job is not runnable - not enqueuing it")
+			continue
+		}
+		nextJLogger.Infof("next job is runnable - enqueuing it")
+		if r.chain.IsSequenceStartJob(nextJob.Id) {
+			// Starting a sequence, so increment sequence try count.
+			r.chain.IncrementSequenceTries(nextJob.Id)
+			seqLogger := r.logger.WithFields(log.Fields{"sequence_id": job.SequenceId, "sequence_try": r.chain.SequenceTries(job.Id)})
+			seqLogger.Info("starting try of sequence")
+		}
+		r.send(nextJob)
+	}
+}
+
+// Send a job to runJobChan. This is very simple, but using this method
+// makes sure every job sent to runJobChan is also added to r.running.
+func (r *RunningChainReaper) send(job proto.Job) {
+	r.running[job.Id] = struct{}{}
+	r.runJobChan <- job
+}
+
+// Finalize determines the final state of the chain and sends it to the Request
+// Manager.
+func (r *RunningChainReaper) Finalize() {
+	_, complete := r.chain.IsDoneRunning()
+	if complete {
+		r.logger.Infof("chain is done, all jobs finished successfully")
+		r.chain.SetState(proto.STATE_COMPLETE)
+	} else {
+		r.logger.Warn("chain is done, some jobs failed")
+		r.chain.SetState(proto.STATE_FAIL)
+	}
+	r.chainRepo.Set(r.chain)
+	if r.sendFinalState() {
+		r.chainRepo.Remove(r.chain.RequestId())
+	}
+}
+
+// -------------------------------------------------------------------------- //
+
+// Job Reaper for chains that are being suspended (stopped to be resumed later).
+type SuspendedChainReaper struct {
+	reaper
+	stopTimeout time.Duration
+}
+
+// Run reaps jobs when they finish running. For each job reaped, if it's...
+// - completed: prepare subsequent jobs (copy jobData).
+// - failed:    prepare a sequence retry.
+// - stopped:   do nothing (job will be retried when chain is resumed).
+func (r *SuspendedChainReaper) Run() {
+	// Wait at most 10 seconds for all running jobs to finish.
+	timeoutChan := make(chan struct{})
+	go func() {
+		time.Sleep(r.stopTimeout)
+		close(timeoutChan)
+	}()
+
+REAPER:
+	for {
+		select {
+		case job := <-r.doneJobChan:
+			delete(r.running, job.Id)
+			r.Reap(job)
+		case <-r.stopChan:
+			close(r.doneStopping)
+			return
+		case <-timeoutChan:
+			break REAPER
+		default:
+			// All jobs have finished running - no more jobs to reap.
+			if len(r.running) == 0 {
+				break REAPER
+			}
+		}
+	}
+
+	r.Finalize()
+	close(r.doneChan)
```

Comment:
Reapers should not be aware of this: doneChan is what stops the `traverser.runJobs` gorountine, but that's owned by the traverser. The traverser and its runJobs is at a higher/different level than the reapers, so the reapers shouldn't be aware of this.

This jives with my earlier comment that we remove the magical timeout from {Stopped|Suspened}Reapder.Run, making the stop or suspend sequence this, all in traverser:

1. Stop RunningReaper
2. Swap in {Stopped|Suspended}Reaper
3. Wait for ^.Run() to return; or timeout and call ^.Stop()
4. defer close(t.doneChan) to stop `traverser.runJobs`

Let me spec several things because this is all related code...